tidy: split dots in filename not the entire path when checking for stray stdout/stderr files

jieyouxu opened the pull request:
I committed a path crime by splitting the entire path on `.`, when I meant to split on the filename. This means that any parent folders which contain `.` will cause tidy failure. Added a regression test so that doesn't happen again.

### Follow-up

- [x] Adjust rustc-dev-guide to document assert on test name not containing dots. https://github.com/rust-lang/rustc-dev-guide/pull/1927

Fixes #121986.

jieyouxu committed:
Split dots in filename, not the entire path

Nilstrieb commented on src/tools/tidy/src/tests_revision_unpaired_stdout_stderr.rs:
```diff
@@ -84,7 +84,9 @@ pub fn check(tests_path: impl AsRef<Path>, bad: &mut bool) {
                 }
             });
 
-            let Some((test_name, _)) = test.to_str().map(|s| s.split_once('.')).flatten() else {
```
You want https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.file_prefix which is sadly unstable

https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.with_extension should work too, and be nicer than... split 😭

jieyouxu replied:
I *think* I actually want `file_stem` semantics here, because I want to keep the test name which is everything but the "rs" extension. This `split_once` doesn't seem right. E.g. `a.b.rs`, the test name is should be `a.b` not `a`. Revisioned output gets additional `.[rev].std{out,err}` after the test name to the best of my knowledge.

jieyouxu replied:
> Could we not use some unstable features?

Tidy requires stable no?

ChrisDenton replied:
Oh 😢

jieyouxu committed:
Use Path::file_stem instead

jieyouxu commented on src/tools/tidy/src/tests_revision_unpaired_stdout_stderr.rs:
```diff
@@ -98,14 +98,20 @@ pub fn check(tests_path: impl AsRef<Path>, bad: &mut bool) {
+                continue;
+            };
+
+            let filename_components = filename.split('.').collect::<Vec<_>>();
+            let [file_prefix, ..] = &filename_components[..] else {
+                continue;
+            };
```
I think this logic here isn't robust either. If a test name is `a.b.rs`, and previously we use `file_stem`, the test name becomes `a.b`. But now this sibling check produces `a` which will silent let `a.b.unknown-revision.stdout` go through because there is no entry in `test_info`. I'll have to think about this part of the logic a bit.

jieyouxu replied:
I'm actually not sure if we *can* properly check here, if we don't assume test name itself to not have any dots.

According to rustc-dev-guide, a test output can take the form

```
<test_name>[.<revision>][.<compare_mode>].<extension>
```

- but `<extension>` can be:
    - `stderr`, `stdout`
    - `run.stderr`, `run.stdout`
    - `64bit.stderr`, `32bit.stderr`
- but `<test_name>` can also have dots if we don't assume it can't have dots.

Maybe it's just better to just say "we assume a test's name don't contain dots or else this tidy check won't catch stray stdout/stderr files for that test"?

workingjubilee replied:
Yes, I think that's necessary to assume for file names. We can assume that of a file name even if we can't assume that for the full path.

onur-ozkan replied:
> we assume a test's name don't contain dots or else this tidy check won't catch stray stdout/stderr files for that test

We may need to add an assertion with a message to ensure that it doesn't happen.

jieyouxu replied:
I'll add an aasert for if the test name contains dots, and update the dev-guide to describe this behavior.

workingjubilee replied:
Probably a good idea. A lot of the test infra assumes that things like e.g. `.64bit.stderr` remain coherent.

jieyouxu committed:
Assert that test names cannot contain dots

This is so that we can catch stray test output files, since test names
without dots can form predictable patterns we can match on.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8152995227/job/22283511904) [(plain)](https://github.com/rust-lang/rust/commit/53f48ddbc7f6a42c2e67f8b16f5868f483821744/checks/22283511904/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=jieyouxu
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_1cfe1e67-a920-4caa-84bc-d717ad451261
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=fix-tidy-unpaired-revision
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_1cfe1e67-a920-4caa-84bc-d717ad451261
GITHUB_REF=refs/pull/121992/merge
GITHUB_REF_NAME=121992/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=1f0faf14cfb19a6a056b2458b717ff0d213a915d
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_1cfe1e67-a920-4caa-84bc-d717ad451261
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_1cfe1e67-a920-4caa-84bc-d717ad451261
GITHUB_TRIGGERING_ACTOR=jieyouxu
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/121992/merge
GITHUB_WORKFLOW_SHA=1f0faf14cfb19a6a056b2458b717ff0d213a915d
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#10 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#10 DONE 0.0s

#11 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#11 0.617   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#11 0.630 Collecting boolean-py==4.0
#11 0.633   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#11 0.646 Collecting chardet==5.1.0
---
#11 3.739 Building wheels for collected packages: reuse
#11 3.740   Building wheel for reuse (pyproject.toml): started
#11 4.079   Building wheel for reuse (pyproject.toml): finished with status 'done'
#11 4.080   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#11 4.080   Stored in directory: /tmp/pip-ephem-wheel-cache-mmeyvu1t/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#11 4.082 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#11 4.104   Attempting uninstall: setuptools
#11 4.105     Found existing installation: setuptools 59.6.0
#11 4.106     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#11 4.106     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#11 4.106     Can't uninstall 'setuptools'. No files were found to uninstall.
#11 4.799 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#11 4.800 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#11 5.326 Collecting virtualenv
#11 5.370   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#11 5.424      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 72.9 MB/s eta 0:00:00
#11 5.473 Collecting platformdirs<5,>=3.9.1
#11 5.477   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#11 5.494 Collecting distlib<1,>=0.3.7
#11 5.497   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#11 5.535 Collecting filelock<4,>=3.12.2
#11 5.553   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#11 5.553   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#11 5.639 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#11 5.805 Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.2.0 virtualenv-20.25.1
#11 DONE 5.9s

#12 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#12 DONE 0.0s
---
DirectMap4k:      198592 kB
DirectMap2M:     7141376 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/5a1e5449c8f4cb6b12b4f64238e3c058767ebf02/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-5a1e5449c8f4cb6b12b4f64238e3c058767ebf02-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished release [optimized] target(s) in 27.47s
##[endgroup]
fmt check
tidy check
thread 'tests_revision_unpaired_stdout_stderr (tests)' panicked at /checkout/src/tools/tidy/src/tests_revision_unpaired_stdout_stderr.rs:91:13:
test name cannot contain dots '.': `/checkout/tests/rustdoc/invalid.crate.name.rs`
thread 'main' panicked at src/tools/tidy/src/main.rs:55:61:
called `Result::unwrap()` on an `Err` value: Any { .. }
Build completed unsuccessfully in 0:00:56
  local time: Tue Mar  5 08:15:27 UTC 2024
```

</details>

jieyouxu commented:
Is there a way to force tidy to rerun on all test files, regardless of if they are modified or not?

onur-ozkan commented:
> > Did you try `--force-rerun` ?
> 
> As in `./x test tidy --force-rerun`? That didn't _seem_ to check all the test files, at least not from the "formatting xxx" messages. Unless it did check, but didn't print anything?

Probably doesn't work then

jieyouxu commented:
> Probably doesn't do anything then

I'll just write an external script to go through all the test files to check if there's any multi-dot test names.

jieyouxu committed:
Update test names to not have dots

onur-ozkan commented:
@bors r+ p=1 (as this fixes a very noisy tidy issue)

bors commented:
:hourglass: Testing commit 247a080b9852e3ab176cca63f5695c8a6b2202cd with merge c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e...
<!-- homu: {"type":"BuildStarted","head_sha":"247a080b9852e3ab176cca63f5695c8a6b2202cd","merge_sha":"c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8156723416/job/22301559102)
Approved by: onur-ozkan
Pushing c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"onur-ozkan","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8156723416/job/22301559102"},"merge_sha":"c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e"} -->

rust-timer commented:
Finished benchmarking commit (c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e): [comparison URL](https://perf.rust-lang.org/compare.html?start=bdde2a80aef587cdbb8eb2d6e295d5c1d05830d9&end=c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=bdde2a80aef587cdbb8eb2d6e295d5c1d05830d9&end=c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bdde2a80aef587cdbb8eb2d6e295d5c1d05830d9&end=c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.2% | [3.2%, 3.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bdde2a80aef587cdbb8eb2d6e295d5c1d05830d9&end=c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=bdde2a80aef587cdbb8eb2d6e295d5c1d05830d9&end=c7beecf3e3cef7a8226a99aec4e4f6bfc114ba8e&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 643.644s -> 644.654s (0.16%)
**Artifact size**: 175.01 MiB -> 175.03 MiB (0.01%)

onur-ozkan commented:
> Is there a way to force tidy to rerun on all test files, regardless of if they are modified or not?

As far as I see we already doing this. Could you tell what made you think that tidy doesn't already do that?
